Samples/Client/SwiftClient/SwiftClient/VideoStreamViewController.swift

```diff
@@ -0,0 +1,640 @@
+            
+            let peerId = Int(peerListDisplayString[startPeerIdIndex..<endPeerIdIndex])!
+            joinPeer(peerId: peerId)
+        }
+    }
+}
+
+extension VideoStreamViewController : RTCPeerConnectionDelegate {
+    public func peerConnection(_ peerConnection: RTCPeerConnection, didChange stateChanged: RTCSignalingState) {}
+    public func peerConnection(_ peerConnection: RTCPeerConnection, didChange newState: RTCIceGatheringState) {}
+    public func peerConnection(_ peerConnection: RTCPeerConnection, didRemove stream: RTCMediaStream) {}
+    public func peerConnectionShouldNegotiate(_ peerConnection: RTCPeerConnection) {}
+    public func peerConnection(_ peerConnection: RTCPeerConnection, didChange newState: RTCIceConnectionState) {}
+    public func peerConnection(_ peerConnection: RTCPeerConnection, didRemove candidates: [RTCIceCandidate]) {}
+    
+    public func peerConnection(_ peerConnection: RTCPeerConnection, didAdd stream: RTCMediaStream) {
+        print("new stream for a remote peer")
+        videoStream = stream
+        if let videoTrack = stream.videoTracks.last {
+            videoTrack.add(renderView)
+            
+            DispatchQueue.main.async {
+                self.view.addSubview(self.renderView)
+            }
+        }
+    }
+    
+    public func peerConnection(_ peerConnection: RTCPeerConnection, didGenerate candidate: RTCIceCandidate) {
+        print("new ice candidate found")
+        let jsonData: [String: String] = [
+            "sdpMLineIndex" : String(candidate.sdpMLineIndex),
+            "sdpMid": candidate.sdpMid!,
+            "candidate": candidate.sdp
+        ]
+        self.sendToPeer(peerId: newIceCandidatePeerId, data: self.convertDictionaryToData(dict: jsonData)!)
+    }
+    
+    public func peerConnection(_ peerConnection: RTCPeerConnection, didOpen dataChannel: RTCDataChannel) {
+        print("didOpenChannel has been called, (js equiv = ondatachannel")
+        inputChannel = dataChannel
+        inputChannel.delegate = self
+    }
+}
+
+extension VideoStreamViewController {
+    func matCreate() -> [CGFloat] {
+        let out: [CGFloat] = [ 1.0, 0.0, 0.0, 0.0,   0.0, 1.0, 0.0, 0.0,  0.0, 0.0, 1.0, 0.0,  0.0, 0.0, 0.0, 1.0]
+        return out
+    }
+    
+    func matClone(a: [CGFloat]) -> [CGFloat] {
+        var out: [CGFloat] = [ 1.0, 0.0, 0.0, 0.0,   0.0, 1.0, 0.0, 0.0,  0.0, 0.0, 1.0, 0.0,  0.0, 0.0, 0.0, 1.0]
+        
+        out[0] = a[0]
+        out[1] = a[1]
+        out[2] = a[2]
+        out[3] = a[3]
+        out[4] = a[4]
+        out[5] = a[5]
+        out[6] = a[6]
+        out[7] = a[7]
+        out[8] = a[8]
+        out[9] = a[9]
+        out[10] = a[10]
+        out[11] = a[11]
+        out[12] = a[12]
+        out[13] = a[13]
+        out[14] = a[14]
+        out[15] = a[15]
+        
+        return out
+    }
+    
+    func matMultiply(a: [CGFloat], b: [CGFloat]) -> [CGFloat] {
```

Comment:
did we borrow this from somewhere? can we give credit if we did?

Reply:
All from mathmatrix.js in the web client, and I will add the appropriate license